stan/math/prim/meta/holder.hpp

```diff
@@ -0,0 +1,274 @@
+template <typename T>
+T holder(T&& arg) {
+  return std::forward<T>(arg);
+}
+
+namespace internal {
+// the function holder_handle_element is also used in holder_cl
+/**
+ * Handles single element (moving rvalues to heap) for construction of
+ * `holder` or `holder_cl` from a functor. For lvalues just sets the `res`
+ * pointer.
+ * @tparam T type of the element
+ * @param a element to handle
+ * @param res resulting pointer to element
+ * @return tuple of pointer allocated on heap (empty).
+ */
+template <typename T>
+auto holder_handle_element(T& a, T*& res) {
+  res = &a;
+  return std::make_tuple();
+}
+
+/**
+ * Handles single element (moving rvalues to heap) for construction of
+ * `holder` or `holder_cl` from a functor. Rvalue is moved to heap and the
+ * pointer to heap memory is assigned to res and returned in a tuple.
+ * @tparam T type of the element
+ * @param a element to handle
+ * @param res resulting pointer to element
+ * @return tuple of pointer allocated on heap (containing single pointer).
+ */
+template <typename T>
+auto holder_handle_element(std::remove_reference_t<T>&& a, T*& res) {
+  res = new T(std::move(a));
+  return std::make_tuple(res);
+}
+
+/**
+ * Second step in implementation of construction `holder` from a functor.
+ * @tparam T type of the result expression
+ * @tparam Is index sequence for `ptrs`
+ * @tparam Args types of pointes to heap
+ * @param expr result expression
+ * @param ptrs pointers to heap that need to be released when the expression is
+ * destructed
+ * @return `holder` referencing given expression
+ */
+template <typename T, std::size_t... Is, typename... Args>
+auto make_holder_impl_step2(T&& expr, std::index_sequence<Is...>,
+                            const std::tuple<Args*...>& ptrs) {
+  return holder(std::forward<T>(expr), std::get<Is>(ptrs)...);
+}
+
+/**
+ * First step in implementation of construction `holder` from a functor.
+ * @tparam T type of the functor
+ * @tparam Is index sequence for `args`
+ * @tparam Args types of arguments
+ * @param func functor
+ * @param args arguments for the functor
+ * @return `holder` referencing given expression
+ */
+template <typename T, std::size_t... Is, typename... Args>
+auto make_holder_impl_step1(const T& func, std::index_sequence<Is...>,
+                            Args&&... args) {
+  std::tuple<std::remove_reference_t<Args>*...> res;
+  auto ptrs = std::tuple_cat(
+      holder_handle_element(std::forward<Args>(args), std::get<Is>(res))...);
+  return make_holder_impl_step2(
+      func(*std::get<Is>(res)...),
+      std::make_index_sequence<std::tuple_size<decltype(ptrs)>::value>(), ptrs);
+}
+
+}  // namespace internal
+
+/**
+ * Constructs an expression from given arguments using given functor.
+ * This is similar to calling the functor with given arguments. Except that any
+ * rvalue argument will be moved to heap first. The arguments moved to heap are
+ * deleted once the expression is destructed.
```

Comment:
I'm getting confused on who is in charge of deleting this

Reply:
Holder object is.

Reply:
Where I'm getting confused in here

```cpp
template <typename T, require_t<std::is_rvalue_reference<T&&>>* = nullptr>
auto holder_handle_element(T&& a, T*& res) {
  res = new T(std::move(a));
  return std::make_tuple(res);
}
```

Where is res deleted at? Is it handled somewhere in Eigen?

Reply:
First this is only used in holder (and its opencl version). So any pointer allocated here ends passed into holder. Holder constructs `std::unique_ptr` around it (see line 65) and `std::unique_ptr` deletes it when its destructor is called.

Reply:
Aight cool!